### PR TITLE
[IBD patch 4] Deprioritise slower downloaders

### PR DESCRIPTION
**Pull request by ptschip**
Here we adjust the max blocks in flight per peer based on it's historical download response times.  This way we ask for more blocks from faster peers.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -7530,30 +7530,35 @@ bool SendMessages(CNode *pto)
...
-                    //  pindex->nHeight, pto->GetLogName());
+                    //     pindex->nHeight, pto->GetLogName());
                 }
             }
+            if (!vGetBlocks.empty())
+            {
+                if (!IsInitialBlockDownload())
+                    requester.AskFor(vGetBlocks, pto);
+                else
+                    requester.AskForDuringIBD(vGetBlocks, pto);
```
Comment (sickpig) [inline src/main.cpp:7304]: wonder if we could have only one `AskFor` to which we pass another bool par to indicate if we are in the IBD phase or not

File: src/unlimited.cpp [hunk 2]
```diff
@@ -1246,7 +1247,18 @@ void IsInitialBlockDownloadInit()
...
     if (!state)
+    {
+        if (!fInitialSyncComplete)
+        {
+            // As we cross the boundary for the first time from IsInitialBlockDownload to !IsInitialBlockDownload, we
+            // want to increment the max header reqs.  What this does is allow us to go back and check more peers to
+            // see if in fact we have the full set of headers; It was possible that during the initial header download
+            // we didn't get everything, either because the peer dropped off or because they fed us an imcomplete set.
+            nMaxHeaderReqsDuringIBD += 3;
+            LOGA("IBD complete - increase max header reqquests to check if we have the full set of headers\n");
```
Comment (sickpig) [inline src/unlimited.cpp]: s/reqquests/request

File: src/net.cpp [hunk 3]
```diff
@@ -1745,7 +1745,8 @@ void ThreadOpenConnections()
         int nOutbound = 0;
         int nThinBlockCapable = 0;
         set<vector<unsigned char> > setConnected;
-        CNode *ptemp = nullptr;
+        CNode *ptemp1 = nullptr;
```
Comment (awemany) [inline src/net.cpp]: ptemp1 and ptemp2 are pretty generic variable names, making this hard to follow :)

**Comment by ptschip** [comment]
thank you ... done.

**Comment by ptschip** [comment]
good idea, having trouble myself following it  ... done.

**Comment by sickpig** [comment]
I've tested the patch  bootstrapping a full nodes multiple times from scratch (modulo the last commit related to `cs_objDownloader` lock) and the IBD time is on par with Core taking into consideration the difference in size of the 2 block chains. Before these set of PRs BU IBD time was significantly higher.

**Comment by ptschip** [comment]
@gandrewstone @sickpig 

Test Results:

Completed 2 full syncs running on Windows 10,   4core i7 8th gen, 16GB RAM storing to SSD, with a wired 150Mbit internet connection and 10GB dbcache.

Ran two tests, 1 a BU node running IBD patch 4, the other ABC node v16.2

BU took 3hr 4  min, ABC 3 hr 23min.  The results are interesting in that they point to even further possible improvements in BU.  The ABC node , although slower overall, really blew away the BU node in the first part of the block chain...after 20 mins, BU was at 245K blocks wheres ABC was already at 265K...however ABC lost all that ground and some in the latter half of the sync.  My observation/analysis shows that it was PV that allowed BU to make up the ground.  The ABC node suffered in the latter part of the sync and could be observed having very large swings in bandwidth use, wheras the BU node was smooth and steady.  I believe this is the result of PV which allows the BU node to ask for and download blocks while it is currently connecting the next block.  So why did ABC do so well in the beginning?  I think it has something to do with the script cache they have which we haven't implemented yet.  I think once we do that  we should be able to shave off another 10 to 20 mins of sync time...once these set of changes are ready and merged I'll work on bringing over the script caching.

**Comment by ptschip** [comment]
@gandrewstone I got the last bug that was causing an intermittent problem.  It was MarkBlocksAsInFlight() which has to be called "BEFORE" we send the block requests , or else we can end up getting blocks back before the block is marked in flight causing a crash.   I think this code is ready for final review.

**Comment by ptschip** [comment]
looks like there is still a bug in here somewhere....intermittent python test failures but only happen on Travis which makes it difficult to debug.

**Comment by sickpig** [comment]
@ptschip I rekick the failing travis job to see if the failure is spurious.

**Comment by ptschip** [comment]
I think this is ready, re-worked this PR and did some cleanup...no issues with travis runs.

Ran a full sync on  e9fb183  forgot to plugin my ethernet cable so this one was over wireless but still was very good...only 4 minutes slower than the previous run, coming in at just under 3hrs and 8 minutes.  Previous run on a wired connection was 3 hrs 4 mins.

File: src/main.cpp (hunk 1 above)
Comment (ptschip) [reply src/main.cpp:7304]: @sickpig I made your requested changed but I've since changed it back.  While in general you are right to suggest this, in this particular case it doesn't work very well.  The problem is that the AskFor() method is also used for getting transactions also; this then makes the code somewhat confusing when I inject code specific for IBD, but in addition we have to then make AskFor() recursive.  I found after I wrote the fix you requested that the code would be very hard to read and make sense for someone unfamiliar with the code, of and coupled with a recursive call that could potentially be a problem if someone in the future updated the code and made an unintended change.  So I thought better to make the code readable and more safe and leave it as it is with the AskForDuringIBD().

Comment (sickpig) [reply src/main.cpp:7304]: fair enough

**Comment by ptschip** [comment]
@gandrewstone @sickpig I fixed one of these mysterious terminations we get on travis...I thought it was something related to this PR but it turns out not...Still the fix is just a workaround...I think the problem has something to do with RPC connections not getting released right away, so by sleeping for  a second before doing join_network (after doing a getchaintips) seems to prevent the script from hanging.

**Comment by sickpig** [comment]
I tested the PR with the last changes checked in e3b6640 and I got even better performance than the latest test I did last week. The IBD improve significantly going from ~10 hrs to ~7hrs (using `-dbcache=10000`).

File: qa/rpc-tests/getchaintips.py [hunk 4]
```diff
@@ -9,7 +9,7 @@
 # This gives us two tips, verify that it works.
 
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import assert_equal
```
Comment (sickpig) [inline qa/rpc-tests/getchaintips.py]: nit: why import everything?

Comment (ptschip) [reply qa/rpc-tests/getchaintips.py]: because i was feeling momentarily lazy...i'll modify it and just use "time"

**Comment by ptschip** [comment]
One last spurious failure to go...this is the one that happens in excessive.py but it seems to get triggered more frequently in this patch...problem is , this is another one that only happens on travis...yuk,

**Comment by ptschip** [comment]
Still getting random failures...hmmm...

**Comment by sickpig** [comment]
@ptschip are you experiencing random failure on you red machine or on travis?

File: src/requestManager.cpp [hunk 5]
```diff
@@ -164,13 +164,43 @@ void CRequestManager::AskFor(const CInv &obj, CNode *from, unsigned int priority
...
+    // in a better order.
+    LOCK(cs_objDownloader);
+    for (auto &inv : objArray)
     {
-        AskFor(objArray[nInv], from, priority);
+        AskFor(inv, from, priority);
     }
 }
 
+void CRequestManager::AskForDuringIBD(const std::vector<CInv> &objArray, CNode *from, unsigned int priority)
```
Comment (sickpig) [inline src/requestManager.cpp]: why not using uint16/32_t as priority type? (if we change here we need to change it also on `AskFor`)

**Comment by ptschip** [comment]
@sickpig at first only travis, but now i've been able to replicate  (at times) on my own local machine.  Something in the last IBD4 patch that has triggered it but it's proving exceedingly difficult to track down.

**Comment by ptschip** [comment]
putting this in WIP while I continue to investigate the cause of the random python test failures for this PR... the previous PR's  IBD1, IBD2 and IBD3 don't have the random failures, only IBD4 has them.

**Comment by ptschip** [comment]
Finally resolved the spurious failure in getchaintips.py

File: src/main.cpp [hunk 6]
```diff
@@ -294,8 +291,11 @@ void FinalizeNode(NodeId nodeid)
     if (state->fSyncStarted)
         nSyncStarted--;
 
-    BOOST_FOREACH (const QueuedBlock &entry, state->vBlocksInFlight)
+    for (const QueuedBlock &entry : state->vBlocksInFlight)
     {
+        printf("erasing map mapblocksinflight entries\n");
```
Comment (gandrewstone) [inline src/main.cpp]: remove

File: src/main.cpp [hunk 7]
```diff
@@ -313,67 +313,79 @@ void FinalizeNode(NodeId nodeid)
...
+            pnode->nMaxBlocksInTransit.store(maxBlocksInTransitPerPeer.value);
         }
         if (blockDownloadWindow.value != 0)
         {
             BLOCK_DOWNLOAD_WINDOW = blockDownloadWindow.value;
         }
-        LOG(THIN, "BLOCK_DOWNLOAD_WINDOW is %d MAX_BLOCKS_IN_TRANSIT_PER_PEER is %d\n", BLOCK_DOWNLOAD_WINDOW,
-            MAX_BLOCKS_IN_TRANSIT_PER_PEER);
+        LOG(THIN, "BLOCK_DOWNLOAD_WINDOW is %d nMaxBlocksInTransit is %d\n", BLOCK_DOWNLOAD_WINDOW,
+            pnode->nMaxBlocksInTransit.load());
```
Comment (gandrewstone) [inline src/main.cpp]: is there a better category for this?  Maybe BLK | THIN

File: src/requestManager.cpp [hunk 8]
```diff
@@ -164,13 +164,43 @@ void CRequestManager::AskFor(const CInv &obj, CNode *from, unsigned int priority
...
+    AskFor(objArray, from, priority);
+
+    // Must lock cs_objDownloader here and before cs_vNodes to maintain proper locking order.
+    LOCK(cs_vNodes);
+    for (CNode *pnode : vNodes)
+    {
+        if (pnode == from)
+            continue;
+
+        AskFor(objArray, pnode, priority);
```
Comment (gandrewstone) [inline src/requestManager.cpp]: it looks here like you are asking for every block from every connected node?

File: src/requestManager.cpp [hunk 9]
```diff
@@ -587,6 +617,31 @@ void CRequestManager::SendRequests()
...
+            {
+                LOCK(cs_main);
+                for (auto &inv : iter.second)
+                {
+                    MarkBlockAsInFlight(iter.first->GetId(), inv.hash, Params().GetConsensus());
+                }
+            }
+            iter.first->PushMessage(NetMsgType::GETDATA, iter.second);
+            ENTER_CRITICAL_SECTION(cs_objDownloader);
+            LOG(REQ, "Sent batched rqst with %d blocks to node %s\n", iter.second.size(), iter.first->GetLogName());
```
Comment (gandrewstone) [inline src/requestManager.cpp]: rqst -> request

File: src/main.cpp (hunk 7 above)
Comment (ptschip) [reply src/main.cpp]: done...added BLK

File: src/requestManager.cpp (hunk 9 above)
Comment (ptschip) [reply src/requestManager.cpp]: done

File: src/requestManager.cpp (hunk 8 above)
Comment (ptschip) [reply src/requestManager.cpp]: yes, we have to do this here.  We need to add a block source for every peer otherwise we can't do re-requests.  But what I should do is (1) only add sources for NETWORK nodes and then (2) perhaps filter for peers that have block availablility , if I can actually do that here, and if I can get the correct block availability ... i'll work on this along with getting the correct block availability for each peer rather than assuming they all have the blocks.

File: src/main.cpp (hunk 6 above)
Comment (ptschip) [reply src/main.cpp]: done

File: src/requestManager.cpp (hunk 8 above)
Comment (sickpig) [reply src/requestManager.cpp]: was wondering, if a node as NODE_NETWORK set and its current height (CH) is higher than our current best chain tip (CT), what is the reason why it wouldn't be able to provide to us any blocks with height H belonging to the (CT, CH]?

**Comment by sickpig** [comment]
travis failure is due to a formatting issue not a tech one

```c++
--- main.cpp
+++ main.cpp_formatted
@@ -362,7 +362,7 @@
                 pnode->nMaxBlocksInTransit.store(16);
             }
 
-            LOG(THIN|BLK, "Average block response time is %.2f seconds\n", pnode->nAvgBlkResponseTime);
+            LOG(THIN | BLK, "Average block response time is %.2f seconds\n", pnode->nAvgBlkResponseTime);
         }
 
         // if there are no blocks in flight then ask for a few more blocks
@@ -377,7 +377,7 @@
         {
             BLOCK_DOWNLOAD_WINDOW = blockDownloadWindow.value;
         }
-        LOG(THIN|BLK, "BLOCK_DOWNLOAD_WINDOW is %d nMaxBlocksInTransit is %d\n", BLOCK_DOWNLOAD_WINDOW,
+        LOG(THIN | BLK, "BLOCK_DOWNLOAD_WINDOW is %d nMaxBlocksInTransit is %d\n", BLOCK_DOWNLOAD_WINDOW,
             pnode->nMaxBlocksInTransit.load());
 
         if (IsChainNearlySyncd())
```

**Comment by ptschip** [comment]
@sickpig  in the normal course of operation when the node has been fully
syncd, we rely on headers arriving from the next mined block to tell us
about the peers block availability.  So block is mined, block propagates
to the peers we are connected to, then those peers announce the block to
us via headers...when we receive the header we do an
UpdateBLockAvailability for that peer and only for the peer we received
the header from...becasue we now know they have the block.  And once we
update their block availablity we then do an AskFor() for that block
(you can see this in the p2p message NetMessage:HEADER section).   Also
very important to keep in mind is that AskFor() doesn't actually ask for
the block, maybe it's a little misleading, askfor, adds this block hash
to the list of sources that the request manager can download from.  So
if this is the first peer, then we do indeed ask for the block...if
we've already requested the block then this second peer becomes a backup
source in case the first fails to give us the block.

OK, now with IBD it's not quite the same because we don't get the header
from every peer.  We only download the headers from one peer (or in the
case of BU right now we do three)... so now we only have updated block
availablity for 3 peers until we receive headers for the next mined
block which could be in an hour or 2 or could be in a minute, but we
can't wait, need them right now.  So that's why I added the code to
assume that every peer we connected to has the blocks we need.  However,
as Andrew got me thinking about it, I think we can improve that here by
first only doing AskFor() from peers that have blockavailability (even
though we assume they do anyway) and then while we're downloading
headers from those 3 initial peers we can also request a header
ocasssionally from every other peer to correctly update everybodies
block availability.

**Comment by sickpig** [comment]
>   So that's why I added the code to assume that every peer we connected to has the blocks we need.  

during IBD I think this is a fair assumption, given that the peer is signaling NODE_NETWORK., i.e. not a pruned node. We could evaluate if the peer is a "good" peer in terms of bandwidth, but I would expect that any given nodes that:

- signals NODE_NETWORK (not pruned)
- has a chaintip higher than our own (or that it's fully synced) 

even if didn't provide us the header of the block at hand should have the block data. 

> However, as Andrew got me thinking about it, I think we can improve that here by first only doing AskFor() from peers that have block availability (even though we assume they do anyway) and then while we're downloading headers from those 3 initial peers we can also request a header occasionally from every other peer to correctly update everybody block availability.

fair enough.

**Comment by ptschip** [comment]
This last series of commits was really just intended to check the correct block availability in AskForDuringIBD()  however I had to move over ProcessBlockAvailaibility and UpdateBlockAvailability to the request manager...then I thought , why not move over FindNextBlocksToDownload() as we would have the IBD code all in one place.  Then I got a little rambunctious and decided to finish the job and bring over MarblocksAsInFlight and MarkBlockAsReceived() as that would have all the block requesting code finally in the request manager which I've been wanting to do for a year now.  There are no logic changes to any of the moves but it but If it's too much I can back so it out for now, but I still need Processblockavailability/Updateblockavailability anyway as they were giving issues.

**Comment by ptschip** [comment]
I have one more item to work on which is getting the correct blockavailability from each peer as we're requesting the initial set of headers during sync.

**Comment by ptschip** [comment]
I think this last commit is good...doing some testing now.

**Comment by ptschip** [comment]
Testing looks good...sync is the same as before 200K + blocks in the first 10 minutes of sync.  I verified we're downloading from all network peers so it looks like the block availability is updating correctly.  I think this is good for final review...

**Comment by sickpig** [comment]
Going to start looking at the code while doing a test IBD using the latest version of the PR

File: src/main.cpp [hunk 10]
```diff
@@ -6398,6 +6098,26 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
...
+            // of headers. Once the reqeusted header is received then the block availability for this peer will get
+            // updated.
+            if (IsInitialBlockDownload())
+            {
+                LOCK(cs_vNodes);
+                for (CNode *pnode : vNodes)
+                {
+                    if (!pnode->fClient && pnode != pfrom)
+                    {
+                        // We only want one single header so we pass a null for CBlockLocator.
```
Comment (gandrewstone) [inline src/main.cpp]: If the block has already given us its best headers, this PushMessage might be unnecessary.  I am also worried about an infinite request loop.  right now this can't happen because PushMessage only asks for a single header, but this code only triggers if nCount == MAX_HEADERS_RESULTS.  But the fact that this prevents an request loop is a secondary effect that might be missed by a future developer.  We can solve both of these by checking that the node doesn't already have a better header here.  I see that check in UpdateBlockAvailability:

if (state->pindexBestKnownBlock == NULL || it->second->nChainWork >= state->pindexBestKnownBlock->nChainWork)
            state->pindexBestKnownBlock = it->second;

Does it make sense to do a similar check here and not send a GETHEADERS if our best block already has greater work than this one?

File: src/requestManager.cpp [hunk 11]
```diff
@@ -587,6 +667,31 @@ void CRequestManager::SendRequests()
             }
         }
     }
+    // send batched requests if any.
+    if (fBatchBlockRequests && !mapBatchBlockRequests.empty())
+    {
+        for (auto iter : mapBatchBlockRequests)
+        {
+            LEAVE_CRITICAL_SECTION(cs_objDownloader);
```
Comment (gandrewstone) [inline src/requestManager.cpp]: we should be able to put this and the ENTER outside of the for loop

File: src/requestManager.cpp [hunk 12]
```diff
@@ -587,6 +667,31 @@ void CRequestManager::SendRequests()
...
+                {
+                    MarkBlockAsInFlight(iter.first->GetId(), inv.hash, Params().GetConsensus());
+                }
+            }
+            iter.first->PushMessage(NetMsgType::GETDATA, iter.second);
+            ENTER_CRITICAL_SECTION(cs_objDownloader);
+            LOG(REQ, "Sent batched request with %d blocks to node %s\n", iter.second.size(), iter.first->GetLogName());
+        }
+
+        LOCK(cs_vNodes);
```
Comment (gandrewstone) [inline src/requestManager.cpp]: This lock is not be necessary I think.  Since we have ownership of the node, it can't be removed from cs_vNodes

File: src/requestManager.cpp (hunk 11 above)
Comment (ptschip) [reply src/requestManager.cpp]: thanks, ok done...also moved up cs_main...we only need to take cs_main once as well.

File: src/main.cpp (hunk 10 above)
Comment (ptschip) [reply src/main.cpp]: hmmm, yes i think you're right, but the problem is cs_main again and locking order...will have to create a copy of vnodes, update refs, release ref's etc...it really bugs me that state is dependent on cs_main...shouldn't have to be...i'll look at that at another time, but for now i'll work on your requested change.

Comment (ptschip) [reply src/main.cpp]: @gandrewstone ok, done...
